src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/StringBuilderAppendShouldNotTakeSubstring.Fixer.cs

```diff
@@ -0,0 +1,173 @@
+                    generator.IdentifierName(nameof(string.Length)));
+
+            var startIndexNode = fixComponents.OriginalInnerArguments[0].Value.Syntax;
+
+            var lengthArgument = generator
+                .SubtractExpression(
+                    lengthNode,
+                    startIndexNode);
+
+            var newNode = generator.InvocationExpression(
+                fixComponents.TargetMethod, 
+                fixComponents.StringArgument.Syntax,
+                startIndexNode, 
+                lengthArgument);
+
+            var newRoot = root
+                .ReplaceNode(
+                    nodeToFix, 
+                    newNode);
+            return document.WithSyntaxRoot(newRoot);
+        }
+
+        private struct FixComponents
+        {
+            public FixComponents(
+                IOperation stringArgument, 
+                SyntaxNode targetMethod, 
+                ImmutableArray<IArgumentOperation> originalInnerArguments)
+            {
+                StringArgument = stringArgument;
+                TargetMethod = targetMethod;
+                OriginalInnerArguments = originalInnerArguments;
+            }
+
+            public IOperation StringArgument { get; }
+            public SyntaxNode TargetMethod { get; }
+            public ImmutableArray<IArgumentOperation> OriginalInnerArguments { get; }
+        }
+
+        private static async Task<Document> FixCodeTwoParameters(
+            Document document,
+            SyntaxNode root,
+            SyntaxNode nodeToFix,
+            CancellationToken cancellationToken)
+        {
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var typedNodeToFix = semanticModel.GetOperation(nodeToFix) as IInvocationOperation;
+            var generator = SyntaxGenerator.GetGenerator(document);
+
+            FixComponents fixComponents = GetFixComponents(generator, typedNodeToFix);
+          
+            var newNode = generator.InvocationExpression(
+                fixComponents.TargetMethod,
+                fixComponents.StringArgument.Syntax,
+                fixComponents.OriginalInnerArguments[0].Syntax,
+                fixComponents.OriginalInnerArguments[1].Syntax);
+            var newRoot = root.ReplaceNode(
+                nodeToFix,
+                newNode);
+
+            return document.WithSyntaxRoot(newRoot);
+        }
+
+        private static FixComponents GetFixComponents(
+            SyntaxGenerator generator, 
+            IInvocationOperation typedNodeToFix)
+        {
+            var argumentOperation = typedNodeToFix.Arguments[0].Value;
+            if (argumentOperation is IInvocationOperation stringBuilderAppendInvocationCandidate)
+            {
+                var innerArguments = stringBuilderAppendInvocationCandidate.Arguments;
+                var instance = stringBuilderAppendInvocationCandidate.Instance;
```

Comment:
i'm not certain i see the value in either of these locals.

Reply:
I neither, now as the code works.
Was still from the exploration phase.